### DISCUSSION TITLE
Identify moved code in PR diffs

### DISCUSSION THREAD

**Discussion by samking** [community/community#8573] [Pull Requests]
This is two related issues:

1. Right now, if I rename a file and also change one line in that file, the GitHub pull request diff will show one file deleted and one file added.  It's difficult to find the one line that's changed, which means that a review that could be very easy is very difficult.
2. Right now, if I move a large function in a file from one place to another (or from one file to another file) and change one line, GitHub just shows it as that function being deleted from one place and added to the other place.  It's difficult to find the one line that's changed, which menas that a review that could be very easy is very difficult.

**Comment by jacob-orbiit** [comment]
> 1. Right now, if I rename a file and also change one line in that file, the GitHub pull request diff will show one file deleted and one file added.  It's difficult to find the one line that's changed, which means that a review that could be very easy is very difficult.

This is a problem with how you've made your commit, not how GitHub is displaying it: To help ensure the move is properly labeled as a "move" in the commit itself, use `git mv`. Git also has some secret threshold (i've never been able to find it officially documented) for determining whether a file has been effectively deleted rather than moved (it seems to be somewhere around 25% of the file's contents changed). `git mv` increases that threshold (to somewhere around 33–50%). One might expect `git mv` to override that threshold completely, but it doesn't (git is otherwise incredibly smart, and this seems a rather obvious mistake, so I would guess git has a good reason for it).



> 2\. Right now, if I move a large function in a file from one place to another (or from one file to another file) and change one line, GitHub just shows it as that function being deleted from one place and added to the other place.  It's difficult to find the one line that's changed, which menas that a review that could be very easy is very difficult.

This would be very nice indeed! Although I'm pretty sure git itself doesn't help with that, so it would be quite difficult for GitHub to support.

**Reply by samking** [reply]
Ah, I was probably running into the secret threshold with my first point then (usually, my real examples are files that are more extensively modified). Thanks for clarifying that!

For my second point: I think that other code review tools like Phabricator support this. It would definitely be a new feature and would take time, but so would every feature request. 🤷

**Reply by francisjeanneau** [reply]
git supports coloring moved lines differently:
https://git-scm.com/docs/git-diff#:~:text=Moved%20lines%20of%20code%20are,with%20no%20mode%20is%20given.

Similar issue: https://github.com/orgs/community/discussions/9632

```
--color-moved[=<mode>]
Moved lines of code are colored differently. It can be changed by the diff.colorMoved configuration setting. The <mode> defaults to no if the option is not given and to zebra if the option with no mode is given. The mode must be one of:

no
Moved lines are not highlighted.

default
Is a synonym for zebra. This may change to a more sensible mode in the future.

plain
Any line that is added in one location and was removed in another location will be colored with color.diff.newMoved. Similarly color.diff.oldMoved will be used for removed lines that are added somewhere else in the diff. This mode picks up any moved line, but it is not very useful in a review to determine if a block of code was moved without permutation.

blocks
Blocks of moved text of at least 20 alphanumeric characters are detected greedily. The detected blocks are painted using either the color.diff.(old|new)Moved color. Adjacent blocks cannot be told apart.

zebra
Blocks of moved text are detected as in blocks mode. The blocks are painted using either the color.diff.(old|new)Moved color or color.diff.(old|new)MovedAlternative. The change between the two colors indicates that a new block was detected.

dimmed-zebra
Similar to zebra, but additional dimming of uninteresting parts of moved code is performed. The bordering lines of two adjacent blocks are considered interesting, the rest is uninteresting. dimmed_zebra is a deprecated synonym.
```

**Comment by cloutierjo** [comment]
I would still point out that in pr file might be identified as deleted and new event if in the individual commit it was detected as a move, this make PR review incredibly difficult even if we specifically made a "move only" commit. I think github should make sure to follow the file history in the pr or other multi commit diff view

**Comment by 10kc-emilyk** [comment]
I noticed that github can properly detect a rename in the `blame` view, when each commit with a move preserves the content, but the PR diff doesnt detect this rename. Perhaps the diff view for PRs could use `blame` for detecting and displaying the PR.

**Comment by hsgg** [comment]
I'm also seeing that Github fails to detect the move (I did use `git mv`), but other tools like `gitk` display the diff properly with only a few lines changed.

(In my case I replaced the original file with a smaller one indicating what happened - maybe that confuses Github.)
